### PR TITLE
crimson/os/seastore/rbm: move mkfs to RBMDevice and use device_config_t

### PR DESCRIPTION
**Pull request by myoungwon**
1. cleanups (bitmap layout)
2. move mkfs to RBMDevice
3. use device_config_t in rbm header

https://github.com/ceph/ceph/pull/47969#discussion_r994013349

Signed-off-by: Myoungwon Oh <myoungwon.oh@samsung.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by myoungwon** [comment]
@cyx1231st Can you take a look when you have time to review?

File: src/crimson/os/seastore/random_block_manager/block_rb_manager.cc [hunk 1]
```diff
@@ -105,8 +102,10 @@ BlockRBManager::write_ertr::future<> BlockRBManager::write(
 {
   ceph_assert(device);
   rbm_abs_addr addr = convert_paddr_to_abs_addr(paddr);
-  if (addr > super.end || addr < super.start ||
-      bptr.length() > super.end - super.start) {
+  rbm_abs_addr start = 0;
+  rbm_abs_addr end = device->get_available_size();
+  if (addr > end || addr < start ||
+      bptr.length() > end - start) {
```
[the remark this answers is not in the document]
Comment (myoungwon) [reply src/crimson/os/seastore/random_block_manager/block_rb_manager.cc]: Done.

File: src/crimson/os/seastore/random_block_manager.h [hunk 2]
```diff
@@ -23,24 +23,42 @@
 
 namespace crimson::os::seastore {
 
+struct rbm_metadata_header_t {
+  size_t size = 0;
+  size_t block_size = 0;
+  uint64_t start; // start location of the device
+  uint64_t end;   // end location of the device
```
[the remark this answers is not in the document]
Comment (myoungwon) [reply src/crimson/os/seastore/random_block_manager.h]: Done.

File: src/crimson/os/seastore/random_block_manager/nvme_block_device.cc [hunk 3]
```diff
@@ -18,6 +18,105 @@ namespace {
...
+namespace crimson::os::seastore::random_block_device {
+#include "crimson/os/seastore/logging.h"
+SET_SUBSYS(seastore_device);
+
+RBMDevice::mkfs_ret RBMDevice::mkfs(device_config_t config) {
+  LOG_PREFIX(RBMDevice::mkfs);
+  super.start = 0;
+  super.end = get_available_size();
+  super.block_size = get_block_size();
+  super.size = get_available_size();
```
[the remark this answers is not in the document]
Comment (myoungwon) [reply src/crimson/os/seastore/random_block_manager/nvme_block_device.cc:32]: Added TODO comment---I will add this part along with a real device (not ephemeral) test because it seems that open_device() and check_create_device() are needed. Do you think this part is required for now?

File: src/crimson/os/seastore/random_block_manager/nvme_block_device.cc [hunk 4]
```diff
@@ -12,12 +12,114 @@
...
     return crimson::get_logger(ceph_subsys_seastore_tm);
   }
 }
+namespace crimson::os::seastore::random_block_device {
+#include "crimson/os/seastore/logging.h"
+SET_SUBSYS(seastore_device);
+
+RBMDevice::mkfs_ret RBMDevice::mkfs(device_config_t config) {
+  LOG_PREFIX(RBMDevice::mkfs);
+  super.start = 0;
```
[the remark this answers is not in the document]
Comment (myoungwon) [reply src/crimson/os/seastore/random_block_manager/nvme_block_device.cc:29]: I renamed RBM_START_ADDRESS. CBJournal has a fixed start position (at the beginning of the device). On the other hand, RBM has a variable position because the two modules coexist on a single device. So, we need super.start where the RBM begins.

File: src/crimson/os/seastore/random_block_manager/nvme_block_device.cc [hunk 5]
```diff
@@ -12,12 +12,114 @@
...
+      decode(super_block, p);
+    }
+    catch (ceph::buffer::error& e) {
+      DEBUG("read_rbm_header: unable to decode rbm super block {}",
+	    e.what());
+      return crimson::ct_error::enoent::make();
+    }
+    checksum_t crc = super_block.crc;
+    bufferlist meta_b_header;
+    super_block.crc = 0;
```
[the remark this answers is not in the document]
Comment (myoungwon) [reply src/crimson/os/seastore/random_block_manager/nvme_block_device.cc:99]: Because header's crc calculation was done with 0. super_block.crc here is the value after calculation.

File: src/crimson/os/seastore/random_block_manager/rbm_device.h [hunk 6]
```diff
@@ -81,9 +88,7 @@ class RBMDevice : public Device {
   // LBA Size
   uint64_t block_size = 4096;
```
[the remark this answers is not in the document]
Comment (myoungwon) [reply src/crimson/os/seastore/random_block_manager/rbm_device.h]: Done.

File: src/crimson/os/seastore/random_block_manager/rbm_device.h [hunk 7]
```diff
@@ -94,14 +99,18 @@ class RBMDevice : public Device {
...
   device_id_t get_device_id() const {
-    return device_id;
+    return super.config.spec.id;
   }
   void set_device_id(device_id_t id) {
-    device_id = id;
+    super.config.spec.id = id;
+  }
+
+  void set_block_size(uint64_t bs) {
```
[the remark this answers is not in the document]
Comment (myoungwon) [reply src/crimson/os/seastore/random_block_manager/rbm_device.h:105]: We need to discuss the disk layout. I originally assumed that a device class matches either CBJournal or RBM. But, now a device can contain the both devices. 
The point is that CBJournal has its own layout---not use device.mkfs---and RBM has a RBM header using device.mkfs. It seems that a reasonable way is that there is only a header using device.mkfs. However, in terms of RBM, we need to handle two different headers on a device. 

The design in my mind is that we allow not to use device.mkfs---cbj.mkfs is called in this case--, setting device member variables. 

What do you think?

File: src/crimson/os/seastore/random_block_manager/nvme_block_device.cc (hunk 5 above)
[the remark this answers is not in the document]
Comment (cyx1231st) [reply src/crimson/os/seastore/random_block_manager/nvme_block_device.cc:105]: Ah, right.

File: src/crimson/os/seastore/random_block_manager/nvme_block_device.cc [hunk 8]
```diff
@@ -12,12 +12,114 @@
...
+	super_block.block_size);
+
+    // Do CRC verification only if data protection is not supported.
+    if (is_data_protection_enabled() == false) {
+      if (meta_b_header.crc32c(-1) != crc) {
+	DEBUG("bad crc on super block, expected {} != actual {} ",
+	      meta_b_header.crc32c(-1), crc);
+	return crimson::ct_error::input_output_error::make();
+      }
+    }
```
[the remark this answers is not in the document]
Comment (cyx1231st) [reply src/crimson/os/seastore/random_block_manager/nvme_block_device.cc:119]: isn't it `ceph_assert_always(crc == -1)` in the else case?

File: src/crimson/os/seastore/random_block_manager/rbm_device.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (cyx1231st) [reply src/crimson/os/seastore/random_block_manager/rbm_device.h]: > The point is that CBJournal has its own layout---does not use device.mkfs()---and RBM has an RBM header created by device.mkfs.

For mkfs CBJournal, how about dropping `cbj_header_t::uuid/block_size/size/device_id` and let `RBMDevice::mkfs()` to configure the size of CBJ area for the device? Seems `Journal::open_for_mkfs()` is sufficient to mkfs the `CBJournal` component --- reset and write down the initial `cbj_header_t` after `Device::mkfs()` is done. And the other mkfs logic in the CBJournal, including `CircularBoundedJournal::mkfs_config_t`, can be dropped.

In short, RBMDevice doesn't need to be coupled with the specific CBJournal layout implementation, it only needs to configure/mkfs the Journal area in this device. Is it reasonable to you?

File: src/crimson/os/seastore/random_block_manager/nvme_block_device.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (cyx1231st) [reply src/crimson/os/seastore/random_block_manager/nvme_block_device.cc:29]: Regarding to https://github.com/ceph/ceph/pull/48478#discussion_r995621995, if the journal area always starts at the block-size, probably we only need to remember the journal-size in the device's superblock, replacing the two existing superblock members `rbm_metadata_header_t::start/start_data_area`, they can be inferred instead.

CBJournal can load the journal range from the backing RBDevice after device mkfs/mount.

[the remark this answers is not in the document]
Comment (myoungwon) [reply src/crimson/os/seastore/random_block_manager/nvme_block_device.cc]: Done.

File: src/crimson/os/seastore/random_block_manager/nvme_block_device.cc (hunk 8 above)
[the remark this answers is not in the document]
Comment (myoungwon) [reply src/crimson/os/seastore/random_block_manager/nvme_block_device.cc:117]: Done.

File: src/crimson/os/seastore/random_block_manager/rbm_device.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (myoungwon) [reply src/crimson/os/seastore/random_block_manager/rbm_device.h]: Sound reasonable. updated.

File: src/crimson/os/seastore/journal/circular_bounded_journal.h [hunk 9]
```diff
@@ -22,8 +22,7 @@
 
 namespace crimson::os::seastore::journal {
 
-constexpr rbm_abs_addr CBJOURNAL_START_ADDRESS = 0;
-constexpr uint64_t CBJOURNAL_MAGIC = 0xCCCC;
+constexpr rbm_abs_addr CBJOURNAL_START_ADDRESS = 4096;
```
Comment (cyx1231st) [inline src/crimson/os/seastore/journal/circular_bounded_journal.h:25]: The RBMDevice defines the start of the journal, so load from `RBMDevice::get_journal_start()`.

File: src/test/crimson/seastore/transaction_manager_test_state.h [hunk 10]
```diff
@@ -116,15 +117,18 @@ class EphemeralTestState {
 
   seastar::future<> randomblock_setup()
   {
-    auto config =
-      journal::CircularBoundedJournal::mkfs_config_t::get_default();
     rb_device.reset(new random_block_device::TestMemory(
```
Comment (cyx1231st) [inline src/test/crimson/seastore/transaction_manager_test_state.h:120]: Create a factory method for `TestMemory` (probably better rename to `EphemeralRBMDevice`) that creates the test device by the cbjournal and data area sizes. Probably cleaner to define and use the `TEST_BLOCK_SIZE` in `TestMemory` for all the tests.

File: src/crimson/os/seastore/journal/circular_bounded_journal.cc [hunk 11]
```diff
@@ -31,23 +28,19 @@ CircularBoundedJournal::CircularBoundedJournal(
     const std::string &path)
   : trimmer(trimmer), device(device), path(path) {}
 
-CircularBoundedJournal::mkfs_ret
-CircularBoundedJournal::mkfs(const mkfs_config_t& config)
+CircularBoundedJournal::mkfs_ret CircularBoundedJournal::mkfs()
```
Comment (cyx1231st) [inline src/crimson/os/seastore/journal/circular_bounded_journal.cc:31]: To work with `TransactionManager::mkfs()`, merge into `open_for_mkfs()`.

File: src/crimson/os/seastore/transaction_manager.cc [hunk 12]
```diff
@@ -649,8 +649,7 @@ TransactionManagerRef make_transaction_manager(
     roll_start = 0;
   } else {
     // FIXME: get from runtime configration instead of static defaults
-    roll_size = journal::CircularBoundedJournal::mkfs_config_t
-                       ::get_default().total_size;
+    roll_size = journal::DEFAULT_TEST_CBJOURNAL_SIZE;
```
Comment (cyx1231st) [inline src/crimson/os/seastore/transaction_manager.cc:652]: Now we can load roll_start and roll_size from `RBMDevice` instead, which implements the FIXME.

File: src/crimson/os/seastore/journal/circular_bounded_journal.h (hunk 9 above)
Comment (myoungwon) [reply src/crimson/os/seastore/journal/circular_bounded_journal.h]: Done.

File: src/test/crimson/seastore/transaction_manager_test_state.h (hunk 10 above)
Comment (myoungwon) [reply src/test/crimson/seastore/transaction_manager_test_state.h]: Done.

File: src/crimson/os/seastore/journal/circular_bounded_journal.cc (hunk 11 above)
Comment (myoungwon) [reply src/crimson/os/seastore/journal/circular_bounded_journal.cc]: Done.

File: src/crimson/os/seastore/transaction_manager.cc (hunk 12 above)
Comment (myoungwon) [reply src/crimson/os/seastore/transaction_manager.cc]: Done.

File: src/crimson/os/seastore/random_block_manager/rbm_device.h [hunk 13]
```diff
@@ -152,25 +153,48 @@ class RBMDevice : public Device {
...
+  using stat_device_ret =
+    read_ertr::future<seastar::stat_data>;
+  virtual stat_device_ret stat_device() = 0;
+
+  uint64_t get_journal_size() const {
+    return super.journal_size;
+  }
 
-class TestMemory : public RBMDevice {
+  void set_journal_size(uint64_t size) {
```
Comment (cyx1231st) [inline src/crimson/os/seastore/random_block_manager/rbm_device.h:171]: IIUC `RBMDevice::set_journal_size()` is for test, group test interfaces together with `set_device_id()` and move to the very end of the class.

File: src/crimson/os/seastore/journal/circular_bounded_journal.h [hunk 14]
```diff
@@ -237,14 +186,16 @@ class CircularBoundedJournal : public Journal {
     auto rbm_tail = get_rbm_addr(get_dirty_tail());
     return rbm_written_to >= rbm_tail ?
       rbm_written_to - rbm_tail :
-      rbm_written_to + header.size + get_block_size()
+      rbm_written_to + get_total_size() + get_block_size()
```
[the remark this answers is not in the document]
Comment (cyx1231st) [reply src/crimson/os/seastore/journal/circular_bounded_journal.h:189]: If that is correct, `get_total_size()` should be `device->get_journal_size() - 2 * get_block_size()`.

The names are confusing, probably:
* get_start_addr -> get_records_start
* get_total_size -> get_records_total_size
* get_available_size -> get_records_available_size
* get_used_size -> get_records_used_size

And group the related interfaces together.

File: src/crimson/os/seastore/random_block_manager/rbm_device.h (hunk 13 above)
Comment (myoungwon) [reply src/crimson/os/seastore/random_block_manager/rbm_device.h]: Done.

File: src/crimson/os/seastore/journal/circular_bounded_journal.h (hunk 14 above)
[the remark this answers is not in the document]
Comment (cyx1231st) [reply src/crimson/os/seastore/journal/circular_bounded_journal.h]: > And `get_records_total_size = records_area_size - block_size` because a block is reserved in the records_area to avoid journal_tail and written_to to overlap each other.
>
> If that is correct, get_records_total_size() should be `device->get_journal_size() - 2 * get_block_size()`.

@myoungwon seems not fully addressed^

File: src/crimson/os/seastore/journal/circular_bounded_journal.h [hunk 15]
```diff
@@ -284,20 +215,68 @@ class CircularBoundedJournal : public Journal {
...
+  /* 
+    Size-related interfaces
+     +---------------------------------------------------------+
+     |   header      | record | record | record | record | ... | 
+     +---------------------------------------------------------+
+     ^               ^                                         ^
+     |               |                                         |
+   get_journal_start |                                     get_journal_end
+              get_records_start
+                     <-------- get_records_total_size --------->
```
Comment (cyx1231st) [inline src/crimson/os/seastore/journal/circular_bounded_journal.h:239]: More precisely, get_records_total_size + block_size

**Review by cyx1231st — APPROVED** [review]
LGTM! One comment left.

File: src/crimson/os/seastore/journal/circular_bounded_journal.h (hunk 15 above)
Comment (myoungwon) [reply src/crimson/os/seastore/journal/circular_bounded_journal.h]: Done.

**Comment by myoungwon** [comment]
@cyx1231st Any other comments?
